src/TransitionReplace/index.jsx

```diff
@@ -0,0 +1,176 @@
+  className: PropTypes.string,
+  onChildEnter: PropTypes.func,
+  onChildEntering: PropTypes.func,
+  onChildEntered: PropTypes.func,
+  onChildExit: PropTypes.func,
+  onChildExiting: PropTypes.func,
+  onChildExited: PropTypes.func,
+  transitionStyles: PropTypes.shape({
+    entering: PropTypes.object,
+    entered: PropTypes.object,
+    exiting: PropTypes.object,
+    exited: PropTypes.object,
+  }),
+  transitionClassNames: PropTypes.string,
+};
+
+const defaultProps = {
+  children: undefined,
+  enterDuration: 300,
+  exitDuration: 300,
+  className: undefined,
+  onChildEnter: undefined,
+  onChildEntering: undefined,
+  onChildEntered: undefined,
+  onChildExit: undefined,
+  onChildExiting: undefined,
+  onChildExited: undefined,
+  transitionStyles: {},
+  transitionClassNames: 'pgn__transition-replace',
+};
+
+
+class TransitionReplace extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = { height: null };
+
+    this.onChildEnter = this.onChildEnter.bind(this);
+    this.onChildEntering = this.onChildEntering.bind(this);
+    this.onChildEntered = this.onChildEntered.bind(this);
+    this.onChildExit = this.onChildExit.bind(this);
+    this.onChildExiting = this.onChildExiting.bind(this);
+    this.onChildExited = this.onChildExited.bind(this);
+  }
+
+  // Transition events are fired in this order:
+  //
+  // onEnter > onEntering > onEntered
+  // onExit  > onExiting  > onExited
+  //
+  // Keep in mind that we always have two transitions happening
+  // both the entering and leaving children
+  //
+  // We set the container height (for animation) in this order:
+  //
+  // 1. onChildExit         (explicitly set the height to match the current current)
+  // 2. onChildEntering     (set the height to match the new content)
+  // 3. onChildExited       (reset the height to null)
+
+  onChildEnter(htmlNode) {
+    if (this.props.onChildEnter) this.props.onChildEnter(htmlNode);
+  }
+
+  onChildEntering(htmlNode) {
+    this.setState({ height: htmlNode.offsetHeight });
+    if (this.props.onChildEntering) this.props.onChildEntering(htmlNode);
+  }
+
+  onChildEntered(htmlNode) {
+    this.setState({ height: null });
+    if (this.props.onChildEntered) this.props.onChildEntered(htmlNode);
+  }
+
+  onChildExit(htmlNode) {
+    this.setState({ height: htmlNode.offsetHeight });
+    if (this.props.onChildExit) this.props.onChildExit(htmlNode);
+  }
+
+  onChildExiting(htmlNode) {
+    if (this.props.onChildExiting) this.props.onChildExiting(htmlNode);
+  }
+
+  onChildExited(htmlNode) {
+    this.setState({ height: null });
+    if (this.props.onChildExited) this.props.onChildExited(htmlNode);
+  }
+
+  renderChildTransition(child) {
+    if (!child.key && process.env.NODE_ENV === 'development') {
+      // eslint-disable-next-line no-console
+      console.warn("TransitionReplace: A child is missing a 'key' prop. Key's are required for any child of this component.");
```

Comment:
Key's -> Keys

This answers my question above about putting this info closer to the code.

Reply:
doh thanks!